### PR TITLE
Refactor into modules, use OS-specific cache dir location

### PR DESCRIPTION
**Pull request by cirla**
Take this all with a grain of salt. More like a code review as a PR.

Changes:

1. Refactor client and config stuff into modules and structs
2. Separated CLI and resolved Config repr
3. Move mockall to dev-dependencies
4. Use anyhow with `.context()` for easier error handling with better messages when something goes wrong
5. Add basic logging (you can set env `RUST_LOG=info` to see the Config log line)
6. Moved `test_data` out of `src` dir
7. Generally you want to use a slice instead of a &Vec, e.g. `&[ItemMapping]`
8. When you need to accept either a `&str` or `String` you can use a generic with restriction that it implements `AsRef<str>` which will match both of those types, and `.as_ref()` will get you the underlying `&str` for either.
9. Just showing an example of RPIT (return position impl Trait) with the `impl Iterator<Item = &ItemMapping>` thing.
10. Organize your imports  and run `cargo fmt` you heathen
11. If you're doing `filter(...).map(...)` its usually more efficient to use `filter_map(...)` instead. You can turn a bool into an option with `then()/then_some()`

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -1,173 +1,96 @@
...
-        .filter(|&mapping| mapping.name.to_lowercase().contains(&lower))
-        .map(|mapping| mapping.id))
+use crate::{
+    client::{Client, ItemMapping, PriceResult},
+    config::Config,
+};
+
+fn get_matching_items<S: AsRef<str>>(
+    name: S,
+    mappings: &[ItemMapping],
```
Comment (darakelian) [inline src/main.rs:15]: Is this `&[foo]` syntax saying the type is just something that has an `iter()` method?

File: src/main.rs [hunk 2]
```diff
@@ -1,173 +1,96 @@
...
+) -> impl Iterator<Item = &ItemMapping> {
```
Comment (darakelian) [inline src/main.rs:16]: Why does this have the `impl` keyword before the `Iterator`

File: src/main.rs [hunk 3]
```diff
@@ -1,173 +1,96 @@
...
 fn display_price(name: &String, price: &PriceResult) {
     let locale = SystemLocale::default().unwrap();
-    println!("{} -> high: {}, low: {}", name, price.high.unwrap_or(0).to_formatted_string(&locale), price.low.unwrap_or(0).to_formatted_string(&locale));
+    println!(
+        "{} -> high: {}, low: {}",
+        name,
+        price
+            .high
+            .map(|p| p.to_formatted_string(&locale))
+            .unwrap_or("N/A".into()),
```
Comment (darakelian) [inline src/main.rs:36]: Does `map` here handle the case where `price.high` is none?

File: src/main.rs (hunk 1 above)
Comment (cirla) [reply src/main.rs:15]: No it’s a [slice](https://doc.rust-lang.org/std/primitive.slice.html) which is like a pointer and length to contiguous items in memory. Vec reference can deref to slice

File: src/main.rs (hunk 2 above)
Comment (cirla) [reply src/main.rs:16]: It’s not a concrete return type, it’s saying “the return type will implement the Iterator trait”

File: src/main.rs (hunk 3 above)
Comment (cirla) [reply src/main.rs:36]: https://doc.rust-lang.org/std/option/enum.Option.html#method.map

Map operates on Some values, ignores and returns None values. The unwrap_or provides a default value if None

**Review by darakelian — APPROVED** [review]

**Pull request merged**